### ISSUE TITLE
Crossings

### ISSUE THREAD

**Issue by dabreegster** [a-b-street/osm2streets#99]
We're starting to represent vehicle movements in #67, but crossings for pedestrians/cyclists are neglected so far. This issue is for collecting notes / ideas.

## Use cases

I have two reasons for working on these in the short-term. First, I'd like to add them to the LTN tool. [Page 18 of this TfL doc](https://content.tfl.gov.uk/lcds-chapter2-toolsandtechniques.pdf) defines "porosity" between areas. Making an area comfortable to move around is half the battle; adjacent areas need comfortable connections over main roads. I'd like to show crossings in the LTN tool, calculate porosity, and then let people sketch in new crossings.

Second, one of the types of walking/cycling infrastructure that ATIP will manage is crossings. Extracting these from OSM, transforming into some simple schema, and thinking about web tools to edit / compare will be important. (so CC @Robinlovelace)

Third, the 15m neighbourhood tool currently assumes you can cross the road at every single intersection and generally doesn't have good time penalties for doing so. Having just returned from Taipei where spending 5 minutes to cross diagonally at a large intersection was normal, I really want to model this kind of thing better.

## References

OSM wiki:
- https://wiki.openstreetmap.org/wiki/Crossings
- https://wiki.openstreetmap.org/wiki/Key:crossing
- https://wiki.openstreetmap.org/wiki/Tag:highway%3Dcrossing

Previous issues discussing this:
- https://github.com/a-b-street/abstreet/issues/795
- https://github.com/a-b-street/abstreet/issues/457
- https://github.com/a-b-street/abstreet/issues/485

## Examples

For testing any new code, I have a few examples in different places I'm very familiar with that demonstrate how hard crossing the street is.

Montlake/520: https://overpass-turbo.eu/s/1mnw
![Screenshot from 2022-10-02 12-14-40](https://user-images.githubusercontent.com/1664407/193451131-4aa7669e-747a-482f-b3d4-58bbf41d0428.png)

Taipei main station: https://overpass-turbo.eu/s/1mnx. The query results don't look right to me -- there's an overpass bridge many blocks to the east, and nothing to the west for a while. See https://docs.google.com/presentation/d/1U1risgMhPWzMwfTFh2HgtSANphe2DXhNePZe6btwu0c/edit#slide=id.g156c4830b34_0_13.
![Screenshot from 2022-10-02 12-17-16](https://user-images.githubusercontent.com/1664407/193451222-dc1ed96c-5ba8-44ae-96dd-60377e809609.png)

By Elephant Park -- there's loads of foot traffic trying to cross Walworth, and the lack of crossings between the roundabout and Heygate feels substantial. https://www.openstreetmap.org/#map=18/51.49263/-0.09835

Westminster bridge: https://overpass-turbo.eu/s/1mny. Every time I wind up in the middle of a crowd on the south side of the bridge, I'm amazed how obnoxious it is to cross north
![Screenshot from 2022-10-02 12-26-53](https://user-images.githubusercontent.com/1664407/193451591-dc27aa2e-c777-4b63-b518-f8ad99249a69.png)

**Comment by Robinlovelace** [comment]
Great to see the 'business logic' of this discussed, and likely in the near future implemented, in the modular osm2streets crate. Very happy to be tagged on this issue and happy to get involved!

**Comment by dabreegster** [comment]
Tiny commit above to just plumb along the positions of `highway=crossing` nodes, without any metadata. I'm experimenting over in the LTN tool with showing crossings over main roads:
![Screenshot from 2022-10-05 14-30-54](https://user-images.githubusercontent.com/1664407/194072752-d97cbd24-a672-4415-887c-84ecca6dc507.png)

One next step will be to classify the types of crossings -- signalized/not, with a pedestrian island/not, bike+foot, etc. I'm still wading through a few different resources to come up with an enum list. Will send a PR for that.

Another next step will be to start thinking about crossings and dual carriageways. Something like this doesn't really help us to know crossing the dual carriageway is possible:
![Screenshot from 2022-10-05 14-33-06](https://user-images.githubusercontent.com/1664407/194073249-ad4879ff-80a7-4547-bb11-d8387f23b46f.png)
OSM has a `footway=crossing` way here, currently filtered out by osm2streets. Still thinking about how to incorporate that. Also, when we do things like collapsing simple sausage links, there's often a crossing in the middle that we could maybe infer and/or insert.

**Comment by BudgieInWA** [comment]
When it comes down to it, a crossing is another movement through an intersection. The fact that we have filtered out footways until now is telling I think: A lot of the logic requires us to only consider car movements. After that, we should also consider 
the conflict between pedestrian/bike "Roads" and car Roads, by adding them as more Road network and their own Movements through intersecions too. As long as you extend far enough down the joining roads as you're tracing the boundary of the area you are calling "intersection" then pedestrian movements enter and exit the intersection in the same way that motor traffic does.

Consider a simple crossing in the middle of a stretch of Road. It's an intersection, but it only has two attached Roads, so it's a trivial `Connection` intersection as far as cars are concerned. (If the crossing is unmarked, you might even want to use a trim offset of 0, so that the intersection itself takes up no area, the longitudinal lines can remain unbroken.) If the crossing is marked, the intersection area interrupts the Roads markings (with the trim offset), and draws the pedestrian markings through the crossing area, along the movement that represents the crossing (that has geometry from a `highway=crossing`, or made up from `sidewalk` tags).

[Here is a crossing](https://ideditor.netlify.app/#background=Bing&disable_features=water,pistes,aerialways,past_future&id=w500106599&map=19.95/-32.06321/115.78151) that uses plenty of "intersection" features, like traffic control, offset stop lines, etc. We should aim to support situations as mapped, and worry about guessing pedestrian islands between close by dual carriageway crossing etc, later.

The intersection up the road (which doesn't have the crossings mapped):
![image](https://user-images.githubusercontent.com/3784591/194205791-4e7c6118-59ae-439a-a555-7d2c6a439315.png)

**Comment by dabreegster** [comment]
Agreed that the crossings should contribute to the intersection geometry and tell us how far back to trim roads...

I just pushed a small change introducing two `CrossingTypes` -- signalized or not. This is a massive over-simplification, but for the first application, it'll suffice and give us an end-to-end usable demo to help flesh out some of the complications. And already, those complications arise -- https://www.openstreetmap.org/node/5761147074 is not marked `crossing = traffic_signal` or with `crossing:signals`, but it's right next to https://www.openstreetmap.org/node/9567380266, a traffic signal node. Maybe we're fine with this and can encourage fixing the tagging, or maybe we'll want to apply some heuristics and fix it up.
